### PR TITLE
Add BSP lifecycle notifications

### PR DESCRIPTION
**Pull request by rmaz**
Sends the BSP lifecycle notifications according to the spec. This PR also adds support for reading the index store path from the build server initialize response metadata.

### CODE REVIEW

File: Sources/SKCore/BuildServerBuildSystem.swift [hunk 1]
```diff
@@ -57,6 +57,15 @@ public final class BuildServerBuildSystem {
     }
   }
 
+  deinit {
+    do {
+      _ = try self.buildServer?.sendSync(ShutdownBuild())
```
Comment (benlangmuir) [inline Sources/SKCore/BuildServerBuildSystem.swift]: Does this need to be synchronous? Seems like you could do the error logging and exit notification from inside the completion block of an async request.

File: Sources/SKCore/BuildServerBuildSystem.swift [hunk 2]
```diff
@@ -78,12 +87,36 @@ public final class BuildServerBuildSystem {
...
+      self.indexStorePath = AbsolutePath(indexStorePath, relativeTo: self.projectRoot)
+    }
     self.buildServer = buildServer
     self.handler = handler
   }
 }
 
+private func readReponseDataKey(data: LSPAny?, key: String) -> String? {
+  switch data {
+  case .dictionary(let dataDict):
```
Comment (benlangmuir) [inline Sources/SKCore/BuildServerBuildSystem.swift]: You can simplify this code a bit with `if case` pattern matching.

```swift
if case .dictionary(let dataDict)? = data,
   case .string(let stringVal)? = dataDict[key]
{
  return stringVal
}

return nil
```

**Comment by benlangmuir** [comment]
@swift-ci please test

File: Sources/SKCore/BuildServerBuildSystem.swift (hunk 2 above)
Comment (rmaz) [reply Sources/SKCore/BuildServerBuildSystem.swift]: Was struggling with this one, knew there had to be a swiftier way to do this.

**Comment by benlangmuir** [comment]
@swift-ci please test

**Comment by benlangmuir** [comment]
@swift-ci please test

**Pull request merged**